src/EditorFeatures/CSharp/CompleteStatement/CompleteStatementCommandHandler.cs

```diff
@@ -0,0 +1,425 @@
+                    return !parenthesizedExpression.CloseParenToken.IsMissing;
+
+                case SyntaxKind.BracketedArgumentList:
+                    var bracketedArgumentList = (BracketedArgumentListSyntax)currentNode;
+                    lastDelimiterSpan = bracketedArgumentList.CloseBracketToken.Span;
+                    return !bracketedArgumentList.CloseBracketToken.IsMissing;
+
+                case SyntaxKind.ObjectInitializerExpression:
+                    var initializerExpressionSyntax = (InitializerExpressionSyntax)currentNode;
+                    lastDelimiterSpan = initializerExpressionSyntax.CloseBraceToken.Span;
+                    return !initializerExpressionSyntax.CloseBraceToken.IsMissing;
+
+                default:
+                    // Type of node does not require a closing delimiter
+                    return true;
+            }
+        }
+
+        private static bool ReachedSemicolonSyntax(SyntaxNode currentNode, ISyntaxFactsService syntaxFacts, ref bool finalDelimiterNeedsSemicolon)
+        {
+            finalDelimiterNeedsSemicolon = false;
+
+            if (currentNode.IsKind(SyntaxKind.DoStatement))
+            {
+                finalDelimiterNeedsSemicolon = true;
+                return true;
+            }
+
+            if (syntaxFacts.IsStatement(currentNode)
+                || currentNode.IsKind(SyntaxKind.GetAccessorDeclaration)
+                || currentNode.IsKind(SyntaxKind.SetAccessorDeclaration)
+                || currentNode.IsKind(SyntaxKind.LocalDeclarationStatement)
+                || currentNode.IsKind(SyntaxKind.VariableDeclaration))
+            {
+                return true;
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// To account for the new line character at the end of a line, this returns the previous tokens end  
+        /// </summary>
+        private static int GetEndPosition(SyntaxNode root, int end, SyntaxKind nodeKind)
+        {
+
+            // If "end" is at the end of a line, the token has trailing end of line trivia.
+            // We want to put our cursor before that trivia, so use previous token for placement.
+            var token = root.FindToken(end);
+
+            if (token.SpanStart >= end)
+            {
+                // We found a token following the end position, which means 'end' is not the end of the line
+                return end;
+            }
+
+            if (token.TrailingTrivia.Any(SyntaxKind.EndOfLineTrivia))
+            {
+                return token.TrailingTrivia.Span.Start;
+            }
+
+            return end;
+            //var previousToken = root.FindToken(end).GetPreviousToken();
+            //return previousToken.Span.End;
+        }
+
+        public VSCommanding.CommandState GetCommandState(TypeCharCommandArgs args, Func<VSCommanding.CommandState> nextCommandHandler) => nextCommandHandler();
+
```

Comment:
extra newlines.

Reply:
➡️ I'm OK with this so we can focus on more important issues.